src/misc/useLocalStorageState.ts

```diff
@@ -0,0 +1,26 @@
+import { useEffect, useState } from "react";
+
+export function useLocalStorageState<T>(key: string, defaultValue: T, options?: { serialize?: (value: T) => string; deserialize?: (value: string) => T; }) {
+  const reactState = useState<T>(() => {
+    const savedValue = localStorage.getItem(key);
+    if (savedValue !== null) {
+      if (options?.deserialize) {
+        return options.deserialize(savedValue);
+      }
+      return savedValue as T; // Assuming T is a string
+    }
+    return defaultValue;
+  })
+
+  const [state] = reactState;
+
+  useEffect(() => {
+    const serializedState = options?.serialize ? options.serialize(state)
+      : (typeof state === 'object' && state !== null)
+        ? JSON.stringify(state) // Fallback for objects
+        : String(state); // Fallback for other types
```

Comment:
Don't love the nested tertiary here... Can we at least simplify it along the lines of:

```suggestion
    const serializedState = options?.serialize?(state) || 
      ((typeof state === 'object' && state !== null)
        ? JSON.stringify(state) // Fallback for objects
        : String(state)); // Fallback for other types
```

Or if not, at least:

```suggestion
    const serializedState = options?.serialize ? options.serialize(state) : basicSerialization(state);
```

And move the nested code into a small helper function?